### DISCUSSION TITLE
Solving common linter problems

### DISCUSSION THREAD

**Discussion by mondeja** [simple-icons/simple-icons#4935] [Help with SVGs]
When you upload an icon to simple-icons, a linting process is triggered to ensure that certain quality standards are met. This process can be executed locally using `npm run svglint` (see how to setup your environment locally in [Contributing Guidelines - Testing package locally](https://github.com/simple-icons/simple-icons/blob/develop/CONTRIBUTING.md#testing-package-locally)). The rules are defined in the file [`.svglintrc.js`](https://github.com/simple-icons/simple-icons/blob/develop/.svglintrc.js), using [SVGLint](https://github.com/birjolaxew/svglint/).

If your icon does not fit a rule, the linter will show you a message explaining the error and maybe a suggestion about how to solve the problem. However, if you can't solve it with the provided report, in this guide are explained some in-depth details about how to solve the most common problems, listed by rule name. If you still having the problem after trying these solutions, don't worry, open the pull anyway if you haven't already, as one of the project maintainers will be happy to help you :smile:

## icon-size

Every icon must have 24 pixels of width or height. This ensures that all icons have the same size. Typically, you can resize the icon with your editor **maintaining its proportions**.

<p align="center">
<img src="https://user-images.githubusercontent.com/23049315/106914290-28af3300-6705-11eb-9dae-540e6cc05d8a.png"/>
</p>
<p align="center"><sup>In Inkscape, make sure that you have activated the lock to keep the proportions when resizing. In other software there should be a similar button.</sup></p>

However, some icons, especially those that touch the outer limits of 24 pixels by a curve, could raise this rule even after being appropriately resized in an editor. This happens because different editors uses different algorithms to compute objects sizes. If you are facing this problem, try one of these solutions:

- Slightly increase or decrease the size of the icon in your editor while resizing. For example, if the linter reports that the size is 24.001 of width, resize to 23.999 of width in your editor.
- Increase the number of decimals of precision used to optimize the result using SVGO, keeping in mind that 5 is the maximum allowed.

## icon-centered

All icons should be centered in the canvas 24x24 pixels. You can center the icon using your preferred editor.

<p align="center"><img src="https://user-images.githubusercontent.com/23049315/106926245-323e9800-6711-11eb-8c13-6aeb5e7d5ce8.gif"/></p>
<p align="center"><sup>In Inkscape, go to <b>Object</b> -> <b>Align and distribute</b> or press <kbd>Ctrl</kbd> <kbd>Shift</kbd> <kbd>A</kbd></sup></p>

However, some icons, especially those that touch the outer limits of 24 pixels by a curve, could raise this rule even after being appropriately centered in an editor. If you are facing this problem, try one of these solutions:

- Slightly moves the icon in X or Y axis after center it. For example, if the linter reports that the center in X is 12.001, move it 0.001 toward the left of the screen.
- Increase the number of decimals of precision used to optimize the result using SVGO, keeping in mind that 5 is the maximum allowed.

## innefective-segments

This rule ensures that coordinates that are not doing anything to draw the icon are removed. For example, moving the SVG virtual pencil to `(2, 2)` and after that drawing a line which ends in `(2, 2)`, as `m2, 2L2, 2` path will do.

- This problem can be solved applying an **union** with your preferred editor.

<p align="center"><img src="https://user-images.githubusercontent.com/23049315/106917394-82fdc300-6708-11eb-9745-5099d5d6220a.jpg"/></p>
<p align="center"><sup>In Inkscape, you'll find at <b>Path</b> -> <b>Union</b></sup></p>

## collinear-segments

This rule ensures that there are no coordinates in the path that are drawing rect lines using redundant commands. For example, moving the SVG virtual pencil to `(2, 2)`, after that drawing a line which ends in `(3, 3)` and then drawing a line that ends in `(4, 4)`, as `m2, 2L3, 3L4, 4` path would do. This can be rewritten just drawing a line toward `(4, 4)` directly with `m2, 2L4, 4`, so the command `L3, 3` should be removed.

<p align="center"><img src="https://user-images.githubusercontent.com/23049315/106921372-3fa55380-670c-11eb-8432-8dff2562b338.png" width="280px"/></p>
<p align="center"><sup>Real example of collinear coordinates</sup></p>

To solve this, you can either:

- Apply an **union** with your preferred editor.
- Replace the incorrect coordinates manually in the corresponding segments (this method require some knowledge of SVG paths syntax).

If the problem persists, try replacing failed coordinates by a high number (`999 999` for example) and you will see the affected coordinates in your editor, so you will be able to edit it manually.

**Comment by fanatiicx** [comment]
Sharing a linter reminder / tip from @ericcornelissen [from this comment](https://github.com/simple-icons/simple-icons/pull/5995#issuecomment-864459894): "the letters in the hex value should be uppercase 😉"
